File: worker/src/analysis.js
// 關鍵字熱度與 ORG 共現的 Worker 端計算，與 src/opinion_pipeline/analysis.py 對齊。
// 全部為可重算的字面統計（計數、log、熵），符合「Worker 只做輕量統計」的邊界。
import { AUTO_TERMS, ORG_LEXICON, WATCH_TERMS } from './generated-config.js';

export const HEAT_WEIGHTS = { volume: 0.5, acceleration: 0.33, diversity: 0.17 };
const KEYWORD_WINDOW_MS = 24 * 60 * 60 * 1000;
const TREND_BUCKETS = 24;
const CJK_RUN = /[㐀-鿿]+/g;

const clamp01 = (value) => Math.min(1, Math.max(0, value));
// 中文不受大小寫影響，只需折疊 ASCII；用 toLowerCase 取代 locale-aware 版本以大幅降低 CPU。
const casefold = (value) => String(value || '').toLowerCase();
const searchTextOf = (item) => `${item.title || ''} ${item.excerpt || ''}`;
// 每筆項目的折疊後文字只計算一次（WeakMap 快取，不污染項目本體）。
const _foldCache = new WeakMap();
function foldOf(item) {
  let value = _foldCache.get(item);
  if (value === undefined) {
    value = casefold(searchTextOf(item));
    _foldCache.set(item, value);
  }
  return value;
}

function matchesFolded(haystack, anyOfFolded, excludeFolded) {
  if (excludeFolded.some((term) => term && haystack.includes(term))) return false;
  return anyOfFolded.some((term) => term && haystack.includes(term));
}

function entropyDiversity(shareValues, enabledSourceCount) {
  if (shareValues.length <= 1 || enabledSourceCount <= 1) return 0;
  const entropy = -shareValues.reduce((acc, p) => (p > 0 ? acc + p * Math.log(p) : acc), 0);
  return clamp01(entropy / Math.log(enabledSourceCount));
}

export function extractAutoTerms(items, cfg = AUTO_TERMS, watchTerms = WATCH_TERMS) {
  const minDocs = Math.max(2, cfg.minDocs);
  const minSources = Math.max(1, cfg.minSources);
  const minLength = cfg.minLength;
  const stopwords = cfg.stopwords || [];
  const watchVocab = watchTerms.flatMap((entry) => [entry.display, ...(entry.anyOf || [])]).filter(Boolean);

  const docCount = new Map();
  const gramDocs = new Map();
  const gramSources = new Map();
  items.forEach((item, docId) => {
    const grams = new Set();
    for (const run of item.title.match(CJK_RUN) || []) {
      for (let size = minLength; size <= 6; size += 1) {
        for (let start = 0; start + size <= run.length; start += 1) grams.add(run.slice(start, start + size));
      }
    }
    for (const gram of grams) {
      docCount.set(gram, (docCount.get(gram) || 0) + 1);
      if (!gramDocs.has(gram)) gramDocs.set(gram, new Set());
      gramDocs.get(gram).add(docId);
      if (!gramSources.has(gram)) gramSources.set(gram, new Set());
      gramSources.get(gram).add(item.source);
    }
  });

  const blocked = (gram) =>
    stopwords.some((stop) => gram.includes(stop)) || watchVocab.some((vocab) => gram.includes(vocab) || vocab.includes(gram));
  // 若某個「更長且被封鎖」的詞涵蓋本詞 ≥80% 的文件，本詞只是它的碎片（如「目標」←「目標價」）。
  const dominatedByBlocked = (gram) => {
    const threshold = 0.8 * docCount.get(gram);
    for (const [cand, count] of docCount) {
      if (cand.length > gram.length && cand.includes(gram) && count >= threshold && blocked(cand)) return true;
    }
    return false;
  };
  const eligible = (gram) =>
    (docCount.get(gram) || 0) >= minDocs &&
    (gramSources.get(gram)?.size || 0) >= minSources &&
    !blocked(gram) &&
    !dominatedByBlocked(gram);

  const promote = (gram) => {
    const threshold = 0.8 * docCount.get(gram);
    let best = gram;
    for (const [cand, count] of docCount) {
      if (cand.length === gram.length + 1 && cand.includes(gram) && count >= threshold && eligible(cand)) {
        if (best === gram || count > docCount.get(best)) best = cand;
      }
    }
    return best;
  };

  const redundant = (gram, chosen) =>
    chosen.some((kept) => {
      if (gram.includes(kept) || kept.includes(gram)) return true;
      const a = gramDocs.get(gram);
      const b = gramDocs.get(kept);
      let overlap = 0;
      for (const doc of a) if (b.has(doc)) overlap += 1;
      return overlap >= 0.6 * Math.min(a.size, b.size);
    });

  const ranked = [...docCount.keys()]
    .filter(eligible)
    .sort((a, b) => docCount.get(b) - docCount.get(a) || b.length - a.length || (a < b ? -1 : 1));

  const chosen = [];
  for (const gram of ranked) {
    if (chosen.length >= cfg.maxTerms) break;
    const term = promote(gram);
    if (!redundant(term, chosen)) chosen.push(term);
  }
  return chosen;
}

export function buildKeywords(items, now = Date.now(), enabledSourceCount = 24, watchTerms = WATCH_TERMS) {
  const windowStart = now - KEYWORD_WINDOW_MS;
  const recent = items.filter((item) => Date.parse(item.publishedAt) >= windowStart);
  const autoTerms = extractAutoTerms(recent);

  const definitions = [];
  for (const entry of watchTerms) {
    const display = String(entry.display || '').trim();
    if (!display) continue;
    const anyOf = (entry.anyOf && entry.anyOf.length ? entry.anyOf : [display]).map(String);
    definitions.push({
      id: `watch-${entry.id || display}`,
      term: display,
      kind: 'manual',
      anyOf,
      exclude: (entry.exclude || []).map(String),
      aliases: anyOf.filter((term) => term !== display),
    });
  }
  autoTerms.forEach((term, index) =>
    definitions.push({ id: `auto-${index + 1}`, term, kind: 'auto', anyOf: [term], exclude: [], aliases: [] }),
  );

  const bucketMs = KEYWORD_WINDOW_MS / TREND_BUCKETS;
  const computed = definitions.map((definition) => {
    const anyOf = definition.anyOf.map(casefold);
    const exclude = definition.exclude.map(casefold);
    const matched = recent.filter((item) => matchesFolded(foldOf(item), anyOf, exclude));
    const sourceCounts = {};
    for (const item of matched) sourceCounts[item.source] = (sourceCounts[item.source] || 0) + 1;
    const total = matched.length;
    const share = {};
    if (total) for (const [src, count] of Object.entries(sourceCounts)) share[src] = Math.round((count / total) * 1000) / 1000;
    const buckets = new Array(TREND_BUCKETS).fill(0);
    for (const item of matched) {
      const index = Math.min(TREND_BUCKETS - 1, Math.floor((Date.parse(item.publishedAt) - windowStart) / bucketMs));
      buckets[index] += 1;
    }
    const recent6 = buckets.slice(-6).reduce((a, b) => a + b, 0);
    const previous6 = buckets.slice(-12, -6).reduce((a, b) => a + b, 0);
    let acceleration = 0;
    if (total) {
      const raw = clamp01(0.5 + (recent6 - previous6) / (2 * Math.max(1, recent6, previous6)));
      acceleration = 0.5 + (raw - 0.5) * Math.min(1, total / 10);
    }
    return { definition, matched: total, share, buckets, acceleration };
  });

  const maxMentions = Math.max(0, ...computed.map((entry) => entry.matched));
  const maxBucket = Math.max(0, ...computed.flatMap((entry) => entry.buckets));
  const keywords = computed.map((entry) => {
    const { definition } = entry;
    const total = entry.matched;
    const volume = maxMentions && total ? clamp01(Math.log1p(total) / Math.log1p(maxMentions)) : 0;
    const diversity = entropyDiversity(Object.values(entry.share), enabledSourceCount);
    const heat =
      Math.round(
        100 *
          (HEAT_WEIGHTS.volume * volume + HEAT_WEIGHTS.acceleration * entry.acceleration + HEAT_WEIGHTS.diversity * diversity) *
          10,
      ) / 10;
    const keyword = {
      id: definition.id,
      term: definition.term,
      kind: definition.kind,
      heat,
      mentions24h: total,
      components: {
        volume: Math.round(volume * 1000) / 1000,
        acceleration: Math.round(entry.acceleration * 1000) / 1000,
        diversity: Math.round(diversity * 1000) / 1000,
        weights: { ...HEAT_WEIGHTS },
      },
      sourceShare: entry.share,
      trend: entry.buckets.map((count, index) => ({
        t: new Date(windowStart + index * bucketMs).toISOString(),
        mentions: count,
        heat: maxBucket ? Math.round((100 * count) / maxBucket * 10) / 10 : 0,
      })),
    };
    if (definition.aliases.length) keyword.aliases = definition.aliases;
    return keyword;
  });

  keywords.sort((a, b) => b.heat - a.heat || b.mentions24h - a.mentions24h || (a.term < b.term ? -1 : 1));
  return keywords;
}

const TOPIC_DEFINITIONS = [
  ['finance', '財經與產業', ['台積電', '半導體', '股市', '經濟', '產業']],
  ['weather', '天氣與防災', ['颱風', '豪雨', '氣象', '地震', '防災']],
  ['politics', '政治與公共政策', ['立法院', '立委', '行政院', '總統', '預算', '政黨']],
  ['society', '社會與生活', ['社會', '交通', '醫療', '健康', '教育', '食安']],
  ['world', '國際與兩岸', ['美國', '中國', '國際', '兩岸', '日本', '歐洲']],
];

const TICKER_NOISE = ['盤中速報', '盤後速報', '近5分K', '三大法人買賣超', '融資融券增減'];
const isTickerNoise = (item) => TICKER_NOISE.some((marker) => searchTextOf(item).includes(marker));

export function buildTopics(items) {
  const topics = [];
  for (const [id, label, terms] of TOPIC_DEFINITIONS) {
    const folded = terms.map(casefold);
    const matched = items.filter((item) => {
      const haystack = foldOf(item);
      return folded.some((term) => haystack.includes(term));
    });
    if (!matched.length) continue;
    const clean = matched.filter((item) => !isTickerNoise(item));
    const preferred = clean.length ? clean : matched;
    const summaries = [];
    for (const item of preferred) {
      const text = (item.excerpt || '').trim() || (item.title || '').trim();
      if (text) summaries.push({ text, source: item.source, url: item.url });
      if (summaries.length === 2) break;
    }
    topics.push({
      id,
      label,
      terms: [...terms],
      size: matched.length,
      sentiment: { positive: 0, neutral: 1, negative: 0 },
      summarySentences: summaries,
      articles: preferred.slice(0, 5).map((item) => ({
        title: item.title,
        source: item.source,
        url: item.url,
        publishedAt: item.publishedAt,
      })),
    });
  }
  return topics;
}

const MIN_NODE_MENTIONS = 2;
const MIN_EDGE_WEIGHT = 2;
const MAX_NODES = 30;

export function buildEntities(items, lexicon = ORG_LEXICON) {
  const folded = lexicon.map((entry) => ({ name: entry.name, terms: [entry.name, ...entry.aliases].map(casefold) }));
  const mentions = new Map();
  const pairDocs = new Map();
  for (const item of items) {
    const haystack = foldOf(item);
    const present = folded.filter((entry) => entry.terms.some((term) => haystack.includes(term))).map((entry) => entry.name);
    present.sort();
    for (const name of present) mentions.set(name, (mentions.get(name) || 0) + 1);
    for (let i = 0; i < present.length; i += 1) {
      for (let j = i + 1; j < present.length; j += 1) {
        const key = `${present[i]} ${present[j]}`;
        pairDocs.set(key, (pairDocs.get(key) || 0) + 1);
      }
    }
  }

  const kept = [...mentions.entries()]
    .filter(([, count]) => count >= MIN_NODE_MENTIONS)
    .sort((a, b) => b[1] - a[1] || (a[0] < b[0] ? -1 : 1))
    .slice(0, MAX_NODES)
    .map(([name]) => name);
  const nodeIds = new Map(kept.map((name, index) => [name, `org-${index + 1}`]));
  const nodes = kept.map((name) => ({ id: nodeIds.get(name), name, type: 'ORG', mentions: mentions.get(name) }));
  const edges = [...pairDocs.entries()]
    .map(([key, weight]) => {
      const [left, right] = key.split(' ');
      return { left, right, weight };
    })
    .filter((edge) => edge.weight >= MIN_EDGE_WEIGHT && nodeIds.has(edge.left) && nodeIds.has(edge.right))
    .sort((a, b) => b.weight - a.weight)
    .map((edge) => ({ source: nodeIds.get(edge.left), target: nodeIds.get(edge.right), weight: edge.weight }));
  return { nodes, edges };
}

File: src/opinion_pipeline/analysis.py
"""從真實新聞快照計算關鍵字熱度與 ORG 共現網絡。

所有數字都可由 news-archive 的 items 重算：
- 關鍵字命中＝標題+短摘要的 substring 比對（大小寫不敏感）。
- 熱度公式：NewsHeat = 100 × (0.50·V + 0.33·A + 0.17·D)
  - V 聲量：24 小時命中數 log1p 後除以當期最大值。
  - A 加速度：近 6 小時相對前 6 小時的成長，0.5 為持平。
  - D 來源多樣性：來源分布熵除以 ln(啟用來源數)。
- 共現＝兩個詞典實體出現在同一篇（標題+短摘要）的獨立文件數。
"""
from __future__ import annotations

import math
import re
from datetime import datetime, timedelta, timezone
from pathlib import Path

import yaml

from .models import NormalizedItem

HEAT_WEIGHTS = {"volume": 0.5, "acceleration": 0.33, "diversity": 0.17}
KEYWORD_WINDOW = timedelta(hours=24)
TREND_BUCKETS = 24
_CJK_RUN_RE = re.compile(r"[㐀-鿿]+")


def load_watch_config(path: Path) -> dict:
    config = yaml.safe_load(path.read_text(encoding="utf-8")) or {}
    return {
        "watch_terms": config.get("watch_terms") or [],
        "auto_terms": config.get("auto_terms") or {},
    }


def _matches(search_text: str, any_of: list[str], exclude: list[str]) -> bool:
    haystack = search_text.casefold()
    if any(term.casefold() in haystack for term in exclude if term):
        return False
    return any(term.casefold() in haystack for term in any_of if term)


def extract_auto_terms(
    items: list[NormalizedItem], watch_terms: list[dict], auto_cfg: dict
) -> list[str]:
    """從標題的 CJK n-gram（2–6 字）統計自動熱詞；全程可重算、無模型。"""
    max_terms = int(auto_cfg.get("max_terms", 10))
    min_docs = max(2, int(auto_cfg.get("min_docs", 2)))
    min_length = int(auto_cfg.get("min_length", 2))
    # 單一媒體的固定版型詞（例如節目名、欄目前綴）不具跨媒體熱度意義。
    min_sources = max(1, int(auto_cfg.get("min_sources", 3)))
    stopwords = [str(word) for word in auto_cfg.get("stopwords") or []]
    watch_vocab = [
        str(term) for entry in watch_terms for term in [entry.get("display", ""), *(entry.get("any_of") or [])] if term
    ]

    doc_count: dict[str, int] = {}
    gram_docs: dict[str, set[int]] = {}
    gram_sources: dict[str, set[str]] = {}
    for doc_id, item in enumerate(items):
        grams: set[str] = set()
        for run in _CJK_RUN_RE.findall(item.title):
            for size in range(min_length, 7):
                for start in range(0, len(run) - size + 1):
                    grams.add(run[start : start + size])
        for gram in grams:
            doc_count[gram] = doc_count.get(gram, 0) + 1
            gram_docs.setdefault(gram, set()).add(doc_id)
            gram_sources.setdefault(gram, set()).add(item.source)

    def blocked(gram: str) -> bool:
        if any(stop in gram for stop in stopwords):
            return True
        return any(gram in vocab or vocab in gram for vocab in watch_vocab)

    def dominated_by_blocked(gram: str) -> bool:
        """若某個「更長且被封鎖」的詞涵蓋了本詞 ≥80% 的文件，本詞只是它的碎片（如「目標」←「目標價」）。"""
        threshold = 0.8 * doc_count[gram]
        for cand, count in doc_count.items():
            if len(cand) > len(gram) and gram in cand and count >= threshold and blocked(cand):
                return True
        return False

    def eligible(gram: str) -> bool:
        return (
            doc_count[gram] >= min_docs
            and len(gram_sources[gram]) >= min_sources
            and not blocked(gram)
            and not dominated_by_blocked(gram)
        )

    def promote(gram: str) -> str:
        """碎片升級：若存在文件數相近（≥80%）、僅多一個字的父字串，改用完整詞（例如「巨蛋」→「大巨蛋」）。"""
        threshold = 0.8 * doc_count[gram]
        candidates = [
            cand
            for cand in doc_count
            if gram in cand and len(cand) == len(gram) + 1 and doc_count[cand] >= threshold and eligible(cand)
        ]
        return max(candidates, key=lambda cand: doc_count[cand]) if candidates else gram

    def redundant(gram: str, chosen: list[str]) -> bool:
        """去冗餘：子/父字串、或與已選詞的命中文件高度重疊（同一事件的不同碎片）。"""
        for kept in chosen:
            if gram in kept or kept in gram:
                return True
            overlap = len(gram_docs[gram] & gram_docs[kept])
            if overlap >= 0.6 * min(len(gram_docs[gram]), len(gram_docs[kept])):
                return True
        return False

    ranked = sorted(
        (gram for gram in doc_count if eligible(gram)),
        key=lambda gram: (-doc_count[gram], -len(gram), gram),
    )
    chosen: list[str] = []
    for gram in ranked:
        if len(chosen) >= max_terms:
            break
        term = promote(gram)
        if not redundant(term, chosen):
            chosen.append(term)
    return chosen


def _clamp01(value: float) -> float:
    return min(1.0, max(0.0, value))


def _entropy_diversity(share: dict[str, float], enabled_source_count: int) -> float:
    if len(share) <= 1 or enabled_source_count <= 1:
        return 0.0
    entropy = -sum(p * math.log(p) for p in share.values() if p > 0)
    return _clamp01(entropy / math.log(enabled_source_count))


def build_keywords(
    items: list[NormalizedItem],
    watch_config: dict,
    now: datetime | None = None,
    enabled_source_count: int = 24,
) -> list[dict]:
    now = now or datetime.now(timezone.utc)
    window_start = now - KEYWORD_WINDOW
    recent = [item for item in items if item.published_at >= window_start]

    watch_terms = watch_config.get("watch_terms") or []
    auto_terms = extract_auto_terms(recent, watch_terms, watch_config.get("auto_terms") or {})

    definitions: list[dict] = []
    for entry in watch_terms:
        display = str(entry.get("display", "")).strip()
        if not display:
            continue
        any_of = [str(term) for term in entry.get("any_of") or [display]]
        definitions.append(
            {
                "id": f"watch-{entry.get('id', display)}",
                "term": display,
                "kind": "manual",
                "any_of": any_of,
                "exclude": [str(term) for term in entry.get("exclude") or []],
                "aliases": [term for term in any_of if term != display],
            }
        )
    for index, term in enumerate(auto_terms):
        definitions.append(
            {"id": f"auto-{index + 1}", "term": term, "kind": "auto", "any_of": [term], "exclude": [], "aliases": []}
        )

    bucket_ms = KEYWORD_WINDOW / TREND_BUCKETS
    computed: list[dict] = []
    for definition in definitions:
        matched = [item for item in recent if _matches(item.search_text, definition["any_of"], definition["exclude"])]
        source_counts: dict[str, int] = {}
        for item in matched:
            source_counts[item.source] = source_counts.get(item.source, 0) + 1
        total = len(matched)
        share = {source: round(count / total, 3) for source, count in source_counts.items()} if total else {}
        buckets = [0] * TREND_BUCKETS
        for item in matched:
            index = min(TREND_BUCKETS - 1, int((item.published_at - window_start) / bucket_ms))
            buckets[index] += 1
        recent6 = sum(buckets[-6:])
        previous6 = sum(buckets[-12:-6])
        if total:
            raw_acceleration = _clamp01(0.5 + (recent6 - previous6) / (2 * max(1, recent6, previous6)))
            # 低聲量時向 0.5（持平）收斂，避免 2 篇新聞就被判定為爆量成長。
            acceleration = 0.5 + (raw_acceleration - 0.5) * min(1.0, total / 10)
        else:
            acceleration = 0.0
        computed.append(
            {
                "definition": definition,
                "matched": total,
                "share": share,
                "buckets": buckets,
                "acceleration": acceleration,
            }
        )

    max_mentions = max([entry["matched"] for entry in computed], default=0)
    max_bucket = max((count for entry in computed for count in entry["buckets"]), default=0)
    keywords: list[dict] = []
    for entry in computed:
        definition = entry["definition"]
        total = entry["matched"]
        volume = _clamp01(math.log1p(total) / math.log1p(max_mentions)) if max_mentions and total else 0.0
        diversity = _entropy_diversity(entry["share"], enabled_source_count)
        acceleration = entry["acceleration"]
        heat = round(
            100
            * (
                HEAT_WEIGHTS["volume"] * volume
                + HEAT_WEIGHTS["acceleration"] * acceleration
                + HEAT_WEIGHTS["diversity"] * diversity
            ),
            1,
        )
        keyword = {
            "id": definition["id"],
            "term": definition["term"],
            "kind": definition["kind"],
            "heat": heat,
            "mentions24h": total,
            "components": {
                "volume": round(volume, 3),
                "acceleration": round(acceleration, 3),
                "diversity": round(diversity, 3),
                "weights": dict(HEAT_WEIGHTS),
            },
            "sourceShare": entry["share"],
            "trend": [
                {
                    "t": (window_start + index * bucket_ms).isoformat().replace("+00:00", "Z"),
                    "mentions": count,
                    "heat": round(100 * count / max_bucket, 1) if max_bucket else 0.0,
                }
                for index, count in enumerate(entry["buckets"])
            ],
        }
        if definition["aliases"]:
            keyword["aliases"] = definition["aliases"]
        keywords.append(keyword)

    keywords.sort(key=lambda value: (-value["heat"], -value["mentions24h"], value["term"]))
    return keywords


# ── ORG 共現網絡 ──────────────────────────────────────────────────────────────

MIN_NODE_MENTIONS = 2
MIN_EDGE_WEIGHT = 2
MAX_NODES = 30


def load_entity_lexicon(path: Path) -> list[dict]:
    config = yaml.safe_load(path.read_text(encoding="utf-8")) or {}
    lexicon = []
    for entry in config.get("orgs") or []:
        if isinstance(entry, str):
            lexicon.append({"name": entry, "aliases": []})
        elif isinstance(entry, dict) and entry.get("name"):
            lexicon.append({"name": str(entry["name"]), "aliases": [str(a) for a in entry.get("aliases") or []]})
    return lexicon


def build_entities(items: list[NormalizedItem], lexicon: list[dict]) -> dict:
    """回傳 {nodes, edges}；只統計詞典內 ORG 在同一篇的共現，不做推論。"""
    mentions: dict[str, int] = {}
    pair_docs: dict[tuple[str, str], int] = {}
    for item in items:
        haystack = item.search_text.casefold()
        present = sorted(
            {
                entry["name"]
                for entry in lexicon
                if any(term.casefold() in haystack for term in [entry["name"], *entry["aliases"]])
            }
        )
        for name in present:
            mentions[name] = mentions.get(name, 0) + 1
        for left_index in range(len(present)):
            for right_index in range(left_index + 1, len(present)):
                pair = (present[left_index], present[right_index])
                pair_docs[pair] = pair_docs.get(pair, 0) + 1

    kept = sorted(
        (name for name, count in mentions.items() if count >= MIN_NODE_MENTIONS),
        key=lambda name: (-mentions[name], name),
    )[:MAX_NODES]
    node_ids = {name: f"org-{index + 1}" for index, name in enumerate(kept)}
    nodes = [{"id": node_ids[name], "name": name, "type": "ORG", "mentions": mentions[name]} for name in kept]
    edges = [
        {"source": node_ids[left], "target": node_ids[right], "weight": weight}
        for (left, right), weight in sorted(pair_docs.items(), key=lambda pair: -pair[1])
        if weight >= MIN_EDGE_WEIGHT and left in node_ids and right in node_ids
    ]
    return {"nodes": nodes, "edges": edges}
